werkzeug/formparser.py

```diff
@@ -378,6 +377,226 @@ def validate_boundary(self, boundary):
+            if not line_terminated:
+                raise ValueError('Unexpected end of line in multipart header')
+            if not line:
+                self._headers = Headers(self._headers)
+                return self._start_content()
+            if line[0] in ' \t' and self._headers:
+                key, value = self._headers[-1]
+                self._headers[-1] = (key, value + '\n ' + line[1:])
+            else:
+                parts = line.split(':', 1)
+                if len(parts) == 2:
+                    self._headers.append((parts[0].strip(), parts[1].strip()))
+                else:
+                    raise ValueError('Malformed header')
+            return self._state_headers
+
+        def feed(self, lines):
+            self._output = []
+            s = self._state
+            for line in lines:
+                s = s(line)
+            self._state = s
+            return self._output
+
+    class PartParser(object):
+        def __init__(self, parent):
+            self.parent = parent
+            self._write = None
+            self._in_memory = 0
+            self._guard_memory = parent.max_form_memory_size is not None
+
+        def _feed_one(self, event):
+            ev, data = event
+            p = self.parent
+            if ev == 'begin_file':
+                self._is_file = True
+            elif ev == 'begin_form':
+                self._headers, self._name = data
+                self._container = []
+                self._write = self._container.append
+                self._is_file = False
+
+            if self._is_file:
+                return event
+
+            if ev == 'cont':
+                self._write(data)
+                if self._guard_memory:
+                    self._in_memory += len(data)
+                    if self._in_memory > p.max_form_memory_size:
+                        p.in_memory_threshold_reached(self._in_memory)
+            elif ev == 'end':
+                part_charset = p.get_part_charset(self._headers)
+                return ('form',
+                        (self._name, b''.join(self._container).decode(
+                            part_charset, p.errors)))
+
+        def feed(self, events):
+            rv = []
+            for event in events:
+                v = self._feed_one(event)
+                if v is not None:
+                    rv.append(v)
+            return rv
+
+    class FileWriter(object):
+        def __init__(self, parent, content_length):
+            self.parent = parent
+            self.content_length = content_length
+            self._file = None
+
+        def _feed_one(self, event):
+            ev, data = event
+            if ev == 'begin_file':
+                self._headers, self._name, filename = data
+                self._filename, self._file = self.parent.start_file_streaming(
+                        filename, self._headers, self.content_length)
+            elif ev == 'cont':
+                self._file.write(data)
```

Comment:
Point of caution: In an async context we wouldn't want to have blocking file writes like this.